File: vol1/ch12/LeetCode327-2.py
class Solution:
    def UpdateOne(self,root,L,R,i):
        if L==R:self.ST[root]+=1
        else:
            mid=(L+R)>>1
            if i<=mid:self.UpdateOne(2*root+1,L,mid,i)
            else:self.UpdateOne(2*root+2,mid+1,R,i)
            self.ST[root]=self.ST[2*root+1]+self.ST[2*root+2]
    def QueryRange(self,root,L,R,i,j):
        if i<=L and j>=R:return self.ST[root]
        else:
            mid=(L+R)>>1
            ans=0
            if i<=mid: ans+=self.QueryRange(2*root+1,L,mid,i,j)
            if j>mid: ans+=self.QueryRange(2*root+2,mid+1,R,i,j)
            return ans

    def countRangeSum(self, nums: List[int],lower:int,upper:int)->int:
        n=len(nums)
        presum=[0]*(n+1)            #前缀和数组
        presum[0]=0
        for i in range(1,n+1):
            presum[i]=presum[i-1]+nums[i-1]
        a=[]                        #离散化
        for x in presum:
            a.append(x)
            a.append(x-lower)
            a.append(x-upper)
        hmap={}
        idx=0
        for x in sorted(a):
            if x not in hmap:
                hmap[x]=idx;idx+=1
        self.ST=[0]*(6*idx)
        ans=0
        for x in presum:
            i,j,k=hmap[x-upper],hmap[x-lower],hmap[x]
            ans+=self.QueryRange(0,0,idx-1,i,j)
            self.UpdateOne(0,0,idx-1,k)
        return ans

File: vol1/ch12/LeetCode327-2.cpp
typedef long long LL;
class Solution {
    vector<int> ST;
public:
    void UpdateOne(int root,LL L,LL R,LL i) {
        if (L==R)
            ST[root]+=1;
        else {
            LL mid=(L+R)>>1;
            if (i<=mid)
                UpdateOne(2*root+1,L,mid,i);
            else
                UpdateOne(2*root+2,mid+1,R,i);
            ST[root]=ST[2*root+1]+ST[2*root+2];
        }
    }
    LL QueryRange(int root, LL L, LL R, LL i, LL j) {
        if (i<=L && j>=R)
            return ST[root];
        else {
            LL mid=(L+R)>>1;
            LL ans=0;
            if (i<=mid) ans+=QueryRange(2*root+1,L,mid,i,j);
            if (j>mid) ans+=QueryRange(2*root+2,mid+1,R,i,j);
            return ans;
        }
    }
    int countRangeSum(vector<int>& nums,int lower,int upper) {
        int n=nums.size();
        vector<LL> presum(n+1);         //前缀和数组
        presum[0]=0;
        for(int i=1;i<=n;i++)
            presum[i]=presum[i-1]+nums[i-1];
        set<LL> tset;                   //除重
        for (LL x:presum) {
            tset.insert(x);
            tset.insert(x-lower);
            tset.insert(x-upper);
        }
        unordered_map<LL,int> hmap;     //离散化
        int idx=0;
        for(LL x:tset) {
            hmap[x]=idx++;
        }
        ST=vector<int>(6*idx,0);
        int ans=0;
        for(LL x: presum) {
            int i=hmap[x-upper],j=hmap[x-lower];
            ans+=QueryRange(0,0,idx-1,i,j);
            int k=hmap[x];
            UpdateOne(0,0,idx-1,k);
        }
        return ans;
    }
};

